### PR TITLE
feat(rollup-relayer): support codecv1

### PR DESCRIPTION
**Pull request by colinlyguo**
### Purpose or design rationale of this PR

- [x] Finalize hardfork name.

Unit tests:
- [x] DB: Insert and get `codecv1` chunk & batch.
- [x] Propose `codecv1` chunks and batches, test `codecv1` limit (e.g., max blob size, previous common limits).
- [x] Process pending/committed batches.
- [x] Integration test 1: passing a hardfork: both parent and current batches are after hardfork, event listening.
- [x] Integration test 2: passing a hardfork: parent batch is before hardfork, current batch is after hardfork, event listening.

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [x] feat: A new feature

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [x] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change

### CODE REVIEW

File: common/docker/l1geth/genesis.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "berlinBlock": 0,
     "londonBlock": 0,
     "clique": {
-      "period": 3,
+      "period": 1,
```
Comment (Thegaram) [inline common/docker/l1geth/genesis.json]: Isn't this ignored in post-merge networks?

File: common/types/encoding/codecv1/codecv1.go [hunk 2]
```diff
@@ -443,8 +443,45 @@ func (b *DABatch) BlobDataProof() ([]byte, error) {
...
 	return nil, nil, nil
 }
+
+// EstimateChunkL1CommitBlobSize estimates the size of the L1 commit blob for a single chunk.
+func EstimateChunkL1CommitBlobSize(c *encoding.Chunk) (uint64, error) {
+	var dataSize uint64
+	for _, block := range c.Blocks {
+		for _, tx := range block.Transactions {
+			if tx.Type != types.L1MessageTxType {
+				rlpTxData, err := encoding.ConvertTxDataToRLPEncoding(tx)
```
Comment (Thegaram) [inline common/types/encoding/codecv1/codecv1.go:494]: Any performance concerns here?

File: common/types/encoding/codecv1/codecv1.go [hunk 3]
```diff
@@ -443,8 +443,45 @@ func (b *DABatch) BlobDataProof() ([]byte, error) {
...
+				if err != nil {
+					return 0, err
+				}
+				dataSize += uint64(len(rlpTxData))
+			}
+		}
+	}
+	paddedSize := ((dataSize + 30) / 31) * 32
+	totalSize := paddedSize + uint64(2+4*MaxNumChunks) // over-estimate: adding metadata length
```
Comment (Thegaram) [inline common/types/encoding/codecv1/codecv1.go]: Need to add `2+4*MaxNumChunks` to `dataSize` before padding (because it is also padded, i.e. it will be 64 bytes not 62).

File: coordinator/internal/orm/chunk.go [hunk 4]
```diff
@@ -161,9 +161,6 @@ func (o *Chunk) getLatestChunk(ctx context.Context) (*Chunk, error) {
 
 	var latestChunk Chunk
 	if err := db.First(&latestChunk).Error; err != nil {
-		if errors.Is(err, gorm.ErrRecordNotFound) {
```
Comment (Thegaram) [inline coordinator/internal/orm/chunk.go]: Decided to undo this change?

File: rollup/conf/config.json [hunk 5]
```diff
@@ -14,7 +14,6 @@
         "escalate_multiple_num": 2,
         "escalate_multiple_den": 1,
         "max_gas_price": 1000000000000,
-        "max_blob_gas_price": 10000000000000,
```
Comment (Thegaram) [inline rollup/conf/config.json:17]: Why is this removed?

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 6]
```diff
@@ -729,6 +643,239 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
...
+		return nil, nil, fmt.Errorf("failed to get parent batch header: %w", getErr)
+	}
+
+	parentBatchHeader := parentDBBatch.BatchHeader
+	parentBatchHash := common.HexToHash(parentDBBatch.Hash)
+	startBlockNumber := dbChunks[0].StartBlockNumber
+	if startBlockNumber >= r.banachForkHeight { // codecv1
+		batch := &encoding.Batch{
+			Index:                      dbBatch.Index,
+			TotalL1MessagePoppedBefore: dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedBefore + dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedInChunk,
```
Comment (Thegaram) [inline rollup/internal/controller/relayer/l2_relayer.go]: I remember we used `parentBatch.TotalL1MessagePopped` before, did you change it so that we don't need to decode the parent batch?

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 7]
```diff
@@ -729,6 +643,239 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
...
+	parentBatchHash := common.HexToHash(parentDBBatch.Hash)
+
+	dbChunks, err := r.chunkOrm.GetChunksInRange(r.ctx, dbBatch.StartChunkIndex, dbBatch.EndChunkIndex)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch chunks: %w", err)
+	}
+
+	startBlockNumber := dbChunks[0].StartBlockNumber
+	if startBlockNumber >= r.banachForkHeight { // codecv1
+		if withProof { // finalizeBatch4844 with proof.
```
Comment (Thegaram) [inline rollup/internal/controller/relayer/l2_relayer.go]: Can we move the chunk and batch encoding plus the blob data proof outside this if-else?

File: rollup/internal/controller/watcher/batch_proposer.go [hunk 8]
```diff
@@ -170,124 +188,82 @@ func (p *BatchProposer) proposeBatch() (*encoding.Batch, error) {
...
-		batch.Index = parentDBBatch.Index + 1
-		parentDABatch, err := codecv0.NewDABatchFromBytes(parentDBBatch.BatchHeader)
-		if err != nil {
-			return nil, err
-		}
-		batch.TotalL1MessagePoppedBefore = parentDABatch.TotalL1MessagePopped
-		batch.ParentBatchHash = common.HexToHash(parentDBBatch.Hash)
+	batch.Index = parentDBBatch.Index + 1
+	var parentDABatch *codecv0.DABatch
+	parentDABatch, err = codecv0.NewDABatchFromBytes(parentDBBatch.BatchHeader)
```
Comment (Thegaram) [inline rollup/internal/controller/watcher/batch_proposer.go]: We might need to use `codecv1` based on `parentDBBatch.StartBlockNumber`.

File: rollup/internal/orm/batch.go [hunk 9]
```diff
@@ -226,7 +222,7 @@ func (o *Batch) GetBatchByIndex(ctx context.Context, index uint64) (*Batch, erro
 }
 
 // InsertBatch inserts a new batch into the database.
-func (o *Batch) InsertBatch(ctx context.Context, batch *encoding.Batch, dbTX ...*gorm.DB) (*Batch, error) {
+func (o *Batch) InsertBatch(ctx context.Context, batch *encoding.Batch, useCodecv0 bool, dbTX ...*gorm.DB) (*Batch, error) {
```
Comment (Thegaram) [inline rollup/internal/orm/batch.go]: I think it would a bit more elegant and readable to pass an enum instead of passing a bool.

File: rollup/internal/orm/batch.go [hunk 10]
```diff
@@ -236,94 +232,126 @@ func (o *Batch) InsertBatch(ctx context.Context, batch *encoding.Batch, dbTX ...
...
-	endDAChunkHash, err := endDAChunk.Hash()
-	if err != nil {
-		log.Error("failed to get end DA chunk hash", "index", batch.Index, "total l1 message popped before", totalL1MessagePoppedBeforeEndDAChunk,
-			"parent hash", batch.ParentBatchHash, "number of chunks", numChunks, "err", err)
-		return nil, fmt.Errorf("Batch.InsertBatch error: %w", err)
+	var batchHash common.Hash
+	var batchBytes []byte
+	var startDAChunkHash common.Hash
+	var endDAChunkHash common.Hash
+	if useCodecv0 {
```
Comment (Thegaram) [inline rollup/internal/orm/batch.go]: It would be nice to use some shared interface instead of if-else. But I'm fine with leaving that as future improvement.

**Review by Thegaram — COMMENTED** [review]
Looks good

File: common/docker/l1geth/genesis.json (hunk 1 above)
Comment (colinlyguo) [reply common/docker/l1geth/genesis.json]: Yes. will remove this change. can remove l1geth after this PR.

File: rollup/conf/config.json (hunk 5 above)
Comment (colinlyguo) [reply rollup/conf/config.json:17]: `LegacyTx`L2  sender (and future possible `DynamicFeeTx` L2 sender) do not need this config.

File: coordinator/internal/orm/chunk.go (hunk 4 above)
Comment (colinlyguo) [reply coordinator/internal/orm/chunk.go]: Yes. This would not happen in a non-genesis path, thus `not found` needs to be considered as an error case, instead of handling a `nil` chunk outside.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 8 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: nice catch. changed in [10ad42c](https://github.com/scroll-tech/scroll/pull/1201/commits/10ad42cf4d568f10ca21323dbd8675e42e73faeb)

File: common/types/encoding/codecv1/codecv1.go (hunk 3 above)
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go]: changed in [10ad42c](https://github.com/scroll-tech/scroll/pull/1201/commits/10ad42cf4d568f10ca21323dbd8675e42e73faeb)

File: common/types/encoding/codecv1/codecv1.go (hunk 2 above)
Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:494]: probably, but from experience, only tracing would result in latency.

File: rollup/internal/controller/relayer/l2_relayer.go (hunk 7 above)
Comment (zimpha) [reply rollup/internal/controller/relayer/l2_relayer.go]: I suggest to make functions like `constructFinalizeBatchPayloadV0` and `constructFinalizeBatchPayloadV1`.

Comment (colinlyguo) [reply rollup/internal/controller/relayer/l2_relayer.go]: good suggestion, will change it into 2 functions.

**Comment by colinlyguo** [comment]
@zimpha contract tests failed, probably due to I synced some code to add unit tests.

**Comment by zimpha** [comment]
> @zimpha contract tests failed, probably due to I synced some code to add unit tests.

yes, I renamed some of the files.

File: common/types/encoding/codecv1/codecv1.go (hunk 2 above)
Comment (Thegaram) [reply common/types/encoding/codecv1/codecv1.go:494]: Ok. Let's test and modify later. (We can still make cost estimation incremental, if instead of exposing `chunk.Blocks`, we add `PushBlock` and `PopBlock` methods. We can consider this later.)

Comment (colinlyguo) [reply common/types/encoding/codecv1/codecv1.go:494]: Good idea.

File: rollup/internal/controller/relayer/l2_relayer.go (hunk 6 above)
Comment (colinlyguo) [reply rollup/internal/controller/relayer/l2_relayer.go]: Rechecked. db batch does not contain `TotalL1MessagePopped`, so fetch this info through the db chunk as a walk-through, another approach is to decode the parent batch's encoding.

File: rollup/internal/orm/batch.go (hunk 10 above)
Comment (colinlyguo) [reply rollup/internal/orm/batch.go]: refactored a bit in [eb74b18](https://github.com/scroll-tech/scroll/pull/1201/commits/eb74b183541b02ba07772032ec07abda04281aef)

File: rollup/internal/orm/batch.go (hunk 9 above)
Comment (colinlyguo) [reply rollup/internal/orm/batch.go]: changed to enum in [eb74b18](https://github.com/scroll-tech/scroll/pull/1201/commits/eb74b183541b02ba07772032ec07abda04281aef)

File: rollup/internal/controller/relayer/l2_relayer.go (hunk 7 above)
Comment (colinlyguo) [reply rollup/internal/controller/relayer/l2_relayer.go]: above comments are addressed in [5fcc831](https://github.com/scroll-tech/scroll/pull/1201/commits/5fcc831f512cb8bf2e416f87941a3d79d2dfb28c)

**Comment by colinlyguo** [comment]
> > @zimpha contract tests failed, probably due to I synced some code to add unit tests.
> 
> yes, I renamed some of the files.

Would we merge your PR first to resolve the CI failure in this PR?

**Comment by zimpha** [comment]
> > > @zimpha contract tests failed, probably due to I synced some code to add unit tests.
> > 
> > 
> > yes, I renamed some of the files.
> 
> Would we merge your PR first to resolve the CI failure in this PR?

I think so.

**Comment by zimpha** [comment]
> > > > @zimpha contract tests failed, probably due to I synced some code to add unit tests.
> > > 
> > > 
> > > yes, I renamed some of the files.
> > 
> > 
> > Would we merge your PR first to resolve the CI failure in this PR?
> 
> I think so.

the PR is merged @colinlyguo

**Comment by colinlyguo** [comment]
> > > > > @zimpha contract tests failed, probably due to I synced some code to add unit tests.
> > > > 
> > > > 
> > > > yes, I renamed some of the files.
> > > 
> > > 
> > > Would we merge your PR first to resolve the CI failure in this PR?
> > 
> > 
> > I think so.
> 
> the PR is merged @colinlyguo

Yeah. there is no diff in the contract repo in this PR now.

File: rollup/internal/controller/watcher/batch_proposer.go (hunk 8 above)
Comment (colinlyguo) [reply rollup/internal/controller/watcher/batch_proposer.go]: update: the above commit introduces a bug (and cannot pass the unit tests), which is fixed in later commits.

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 11]
```diff
@@ -729,6 +714,140 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
...
+	if packErr != nil {
+		return nil, fmt.Errorf("failed to pack commitBatch: %w", packErr)
+	}
+	return calldata, nil
+}
+
+func (r *Layer2Relayer) constructCommitBatchPayloadCodecV1(dbBatch *orm.Batch, dbParentBatch *orm.Batch, dbChunks []*orm.Chunk, chunks []*encoding.Chunk) ([]byte, *kzg4844.Blob, error) {
+	batch := &encoding.Batch{
+		Index:                      dbBatch.Index,
+		TotalL1MessagePoppedBefore: dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedBefore + dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedInChunk,
```
Comment (Thegaram) [inline rollup/internal/controller/relayer/l2_relayer.go]: Isn't this popped after?

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 12]
```diff
@@ -729,6 +714,140 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
...
+	if packErr != nil {
+		return nil, fmt.Errorf("failed to pack finalizeBatch: %w", packErr)
+	}
+	return calldata, nil
+}
+
+func (r *Layer2Relayer) constructFinalizeBatchPayloadCodecV1(dbBatch *orm.Batch, dbParentBatch *orm.Batch, dbChunks []*orm.Chunk, chunks []*encoding.Chunk, aggProof *message.BatchProof) ([]byte, error) {
+	batch := &encoding.Batch{
+		Index:                      dbBatch.Index,
+		TotalL1MessagePoppedBefore: dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedBefore + dbChunks[len(dbChunks)-1].TotalL1MessagesPoppedInChunk,
```
Comment (Thegaram) [inline rollup/internal/controller/relayer/l2_relayer.go]: Same

File: rollup/internal/controller/relayer/l2_relayer.go [hunk 13]
```diff
@@ -729,6 +714,140 @@ func (r *Layer2Relayer) handleL2RollupRelayerConfirmLoop(ctx context.Context) {
...
+		)
+		if packErr != nil {
+			return nil, fmt.Errorf("failed to pack finalizeBatchWithProof4844: %w", packErr)
+		}
+		return calldata, nil
+	}
+
+	// finalizeBatch4844 without proof.
+	calldata, packErr := r.l1RollupABI.Pack(
+		"finalizeBatch4844", // Assuming the "to be implemented" bypass function name is finalizeBatch4844.
```
Comment (Thegaram) [inline rollup/internal/controller/relayer/l2_relayer.go]:
```suggestion
		"finalizeBatch4844",
```

File: rollup/internal/controller/relayer/l2_relayer.go (hunk 11 above)
Comment (Thegaram) [reply rollup/internal/controller/relayer/l2_relayer.go]: Fixed in [0d72756](https://github.com/scroll-tech/scroll/pull/1201/commits/0d7275644803acdcfd3a006cd73c44d3e51edf42).

**Review by Thegaram — DISMISSED** [review]

**Review by Thegaram — APPROVED** [review]
